File: src/languages/maltese/maltese.ts
/**
 * Maltese (mt) phonemizer — Malti, the only Semitic language in the Latin alphabet, canonical IPA,
 * Maltese orthography is fairly phonemic, so this is a greedy grapheme scan (the ⟨ie għ⟩
 * digraphs + the silent-letter rules) + final devoicing + regressive voicing assimilation + ⟨n⟩→m before a labial.
 *   - ⟨q⟩→ʔ (glottal stop), ⟨ċ⟩→t͡ʃ, ⟨ġ⟩→d͡ʒ, ⟨ħ⟩→ħ, ⟨x⟩→ʃ, ⟨z⟩→t͡s vs ⟨ż⟩→z.
 *   - ⟨għ⟩ is SILENT (historically /ʕ ɣ/; it lengthens/pharyngealizes an adjacent vowel — folded).
 *   - ⟨h⟩ is SILENT word-medially (lengthens the adjacent vowel: deheb→dɛːp) but → [ħ] WORD-FINAL (fih→fiːħ).
 *   - ⟨ie⟩ → the long [ɪː] (folded to ɪ). Adjacent identical vowels (from a dropped silent ⟨h/għ⟩) collapse.
 * Vowel LENGTH (stress-conditioned, ~half the referee) + stress (unwritten) are folded/deferred.
 */
import type { Phonemizer } from "../../registry.ts";
import { assembleClauses } from "../../core/clauses.ts";
import { hostWordRun, makeNativiser } from "../../core/hostWord.ts";
import { loadManifest } from "../../core/loadManifest.ts";
import { type MalteseNumbers, numberToWords, readDigits } from "./numbers.ts";

interface MalteseDef {
    graphemes: Record<string, string>;
    voicing: { devoice: Record<string, string>; voice: Record<string, string> };
    numbers: MalteseNumbers;
    clausePunctuation: Record<string, string>;
}
const DEF = loadManifest<MalteseDef>(import.meta.url, "maltese.jsonc");
const G = DEF.graphemes;
const DEVOICE = DEF.voicing.devoice;
const VOICE = DEF.voicing.voice;
const CLAUSE_MARK = DEF.clausePunctuation;
const NUM = DEF.numbers;
const VOWELS = new Set([..."aɛɪɔu"]);
const VOICELESS = new Set([...Object.values(DEVOICE), "t͡s", "ʃ", "ħ", "ʔ", "f", "k", "p", "t", "s"]);

/** Scan a lowercased Maltese word into IPA phoneme tokens (digraphs ⟨ie għ⟩ + the ⟨h⟩ silent/[ħ] rule). */
function scan(word: string): string[] {
    const w = [...word.toLowerCase()];
    const toks: string[] = [];
    for (let i = 0; i < w.length; i++) {
        const c = w[i]!, two = c + (w[i + 1] ?? "");
        if (two === "għ") { if (i + 2 >= w.length) toks.push("ħ"); i += 1; continue; } // ⟨għ⟩: [ħ] word-final (biegħ→bɪːħ), else silent
        if (two === "ie") { toks.push("ɪ"); i += 1; continue; } // ⟨ie⟩ → long [ɪː] (ː folded)
        if (c === "h") { if (i === w.length - 1) toks.push("ħ"); continue; } // ⟨h⟩: [ħ] word-final, else silent
        const ph = G[c];
        if (ph !== undefined) toks.push(ph);
    }
    return toks;
}

/** Collapse two adjacent identical vowels (from a dropped silent ⟨h/għ⟩) into one — deheb→dɛ(h)ɛb → dɛb (the long
 *  vowel is folded): Maltese does not write geminate short vowels, so an adjacent pair is a silent-letter artifact. */
function collapseVowels(toks: string[]): void {
    for (let i = toks.length - 1; i > 0; i--) {
        if (toks[i] === toks[i - 1] && VOWELS.has(toks[i]!)) toks.splice(i, 1);
    }
}

/** Word-final geminate DEGEMINATION: a doubled consonant collapses to a single word-finally (Ħadd→ħat, att→at) —
 *  but a MEDIAL geminate is kept (attard→attart, qattus→ʔattus). */
function degeminateFinal(toks: string[]): void {
    const n = toks.length;
    if (n >= 2 && toks[n - 1] === toks[n - 2] && !VOWELS.has(toks[n - 1]!)) toks.pop();
}

// A geminated AFFRICATE surfaces as its STOP + the affricate (⟨ġġ⟩→[dd͡ʒ], ⟨ċċ⟩→[tt͡ʃ], ⟨zz⟩→[tt͡s]).
const AFFRICATE_STOP: Record<string, string> = { "d͡ʒ": "d", "t͡ʃ": "t", "t͡s": "t" };
/** Affricate gemination: two adjacent identical affricates → stop + affricate (mweġġa'→mwɛdd͡ʒa). */
function geminateAffricate(toks: string[]): void {
    for (let i = 0; i < toks.length - 1; i++) {
        if (toks[i]! in AFFRICATE_STOP && toks[i] === toks[i + 1]) toks[i] = AFFRICATE_STOP[toks[i]!]!;
    }
}

/** ⟨n⟩ → m before a labial (b/p/m): ġenb→d͡ʒɛmp. */
function labialNasal(toks: string[]): void {
    for (let i = 0; i < toks.length - 1; i++) {
        if (toks[i] === "n" && ["b", "p", "m"].includes(toks[i + 1]!)) toks[i] = "m";
    }
}

/** Regressive voicing assimilation over obstruent clusters + word-final devoicing (Attard→attart, ħobż→ħɔps). */
function applyVoicing(toks: string[]): void {
    const last = toks.length - 1;
    if (last >= 0 && toks[last]! in DEVOICE) toks[last] = DEVOICE[toks[last]!]!; // word-final devoicing
    for (let k = toks.length - 2; k >= 0; k--) {
        const b = toks[k]!, nb = toks[k + 1]!;
        if (b in DEVOICE && VOICELESS.has(nb)) toks[k] = DEVOICE[b]!; // devoice before a voiceless obstruent
        else if (b in VOICE && nb in DEVOICE) toks[k] = VOICE[b]!; // voice before a voiced obstruent
    }
}

/** Phonemize a single Maltese word to canonical IPA (segmental; length + stress folded/deferred). */
export function phonemizeWord(word: string): string {
    const toks = scan(word);
    collapseVowels(toks);
    degeminateFinal(toks);
    geminateAffricate(toks);
    labialNasal(toks);
    applyVoicing(toks);
    return toks.join("");
}

// A word (Maltese Latin letters incl. ċ ġ ħ ż) / number / punctuation token.
const TOKEN = new RegExp(`(${hostWordRun(["Latin"], "'")})|(\\d+)|([.!?…,;:])`, "gu");

/**
 * This language's OWN inventory — the TOKEN word class as it stood before the widening above, lifted
 * verbatim, so nothing about the orthography is invented here. A token this REJECTS carries a letter the
 * language does not use, i.e. a foreign name. See core/hostWord.ts: this is the INVENTORY question, and it
 * is no longer also deciding where the script boundary falls.
 *
 * ⚠ NOT QUITE VERBATIM: à ò À Ò were REMOVED, because the g2p has no rule for them and DROPPED them outright.
 * The old token class listed them anyway, and the word-level fold hid the mismatch — a word containing one was
 * rejected whole, so everything in it got folded and the letter came out readable by accident. Judging each
 * character on its own exposes the over-claim instead of masking it: `Thérèse` in Romanian read *ˈthrese*, the é
 * gone, because the class promised a rule that did not exist. NATIVE_CLASS is a claim about the G2P, and
 * `test/native-inventory.test.ts` now measures it rather than trusting it.
 */
const NATIVE_CLASS = "[a-zċġħżèìùA-ZĊĠĦŻÈÌÙ']";
const nat = makeNativiser(NATIVE_CLASS, "u");

class MaltesePhonemizer implements Phonemizer {
    text(input: string): string {
        return assembleClauses(input, TOKEN, (m, sink) => {
            if (m[1]) sink.emit(phonemizeWord(nat(m[1])));
            else if (m[2]) {
                // ≤12 digits stays inside the attested range (< 10¹²); longer reads the raw digit string so the
                // Number() conversion can't lose precision or go exponential. See numbers.ts for the sources.
                const words = m[2].length <= 12 ? numberToWords(Number(m[2]), NUM) : readDigits(m[2], NUM);
                // "tnax-il" is one WORD whose hyphen the grapheme scan simply skips (→ ħdaxil), which is the
                // correct Maltese pronunciation of the linker; it is not a token boundary.
                for (const wd of words.split(" ")) sink.emit(phonemizeWord(wd));
            } else if (m[3]) {
                const mk = CLAUSE_MARK[m[3]];
                if (mk) sink.pause(mk);
            }
        });
    }
}

/** Build the Maltese phonemizer (grapheme g2p + silent-letter rules + devoicing; length/stress folded/deferred). */
export function createMaltese(): Phonemizer {
    return new MaltesePhonemizer();
}

File: csharp/Vernacula.Phonemizer/Languages/Maltese/Maltese.cs
/**
 * Maltese (mt) phonemizer — Malti, the only Semitic language in the Latin alphabet, canonical IPA. Maltese
 * orthography is fairly phonemic, so this is a greedy grapheme scan (the ⟨ie għ⟩ digraphs plus the
 * silent-letter rules) + final devoicing + regressive voicing assimilation + ⟨n⟩→m before a labial.
 *   · ⟨q⟩→ʔ (glottal stop), ⟨ċ⟩→t͡ʃ, ⟨ġ⟩→d͡ʒ, ⟨ħ⟩→ħ, ⟨x⟩→ʃ, ⟨z⟩→t͡s vs ⟨ż⟩→z.
 *   · ⟨għ⟩ is SILENT (historically /ʕ ɣ/; it lengthens or pharyngealizes an adjacent vowel — folded), but
 *     surfaces as [ħ] WORD-FINAL (biegħ → bɪːħ).
 *   · ⟨h⟩ is SILENT word-medially (lengthening the adjacent vowel: deheb→dɛːp) but → [ħ] WORD-FINAL (fih→fiːħ).
 *   · ⟨ie⟩ → the long [ɪː] (folded to ɪ). Adjacent identical vowels — from a dropped silent ⟨h/għ⟩ — collapse.
 * Vowel LENGTH (stress-conditioned, about half the referee) and stress (unwritten) are folded and deferred.
 */
using Vernacula.Phonemizer.Core;

namespace Vernacula.Phonemizer.Languages.Maltese;

public sealed class MaltesePhonemizer : ILanguage
{
    private static IReadOnlyDictionary<string, string> G => Manifest.DEF.Graphemes;
    private static IReadOnlyDictionary<string, string> DEVOICE => Manifest.DEF.Voicing.Devoice;
    private static IReadOnlyDictionary<string, string> VOICE => Manifest.DEF.Voicing.Voice;
    private static IReadOnlyDictionary<string, string> CLAUSE_MARK => Manifest.DEF.ClausePunctuation;
    private static MalteseNumbers NUM => Manifest.DEF.Numbers;
    private static readonly IReadOnlySet<string> VOWELS = Ipa.IPA_VOWEL;
    private static readonly IReadOnlySet<string> VOICELESS = new HashSet<string>(
        Manifest.DEF.Voicing.Devoice.Values.Concat(new[] { "t͡s", "ʃ", "ħ", "ʔ", "f", "k", "p", "t", "s" }),
        StringComparer.Ordinal);

    /**
     * Scan a lowercased Maltese word into IPA phoneme tokens (the ⟨ie għ⟩ digraphs plus the ⟨h⟩ rule).
     * ⚠ CODE POINTS — the TS spells this `[...word.toLowerCase()]`.
     */
    private static List<string> Scan(string word)
    {
        var w = Js.CodePoints(Js.ToLowerCase(word));
        var toks = new List<string>();
        for (var i = 0; i < w.Count; i++)
        {
            var c = w[i];
            var two = c + (i + 1 < w.Count ? w[i + 1] : "");
            // ⟨għ⟩: [ħ] word-final (biegħ→bɪːħ), else silent.
            if (two == "għ") { if (i + 2 >= w.Count) toks.Add("ħ"); i += 1; continue; }
            if (two == "ie") { toks.Add("ɪ"); i += 1; continue; } // ⟨ie⟩ → long [ɪː], the ː folded
            if (c == "h") { if (i == w.Count - 1) toks.Add("ħ"); continue; } // [ħ] word-final, else silent
            if (G.TryGetValue(c, out var ph)) toks.Add(ph);
        }
        return toks;
    }

    /** Collapse two adjacent identical vowels (from a dropped silent ⟨h/għ⟩) into one — deheb→dɛ(h)ɛb→dɛb.
     *  Maltese does not write geminate short vowels, so an adjacent pair is a silent-letter artifact. */
    private static void CollapseVowels(List<string> toks)
    {
        for (var i = toks.Count - 1; i > 0; i--)
            if (toks[i] == toks[i - 1] && VOWELS.Contains(toks[i])) toks.RemoveAt(i);
    }

    /** Word-final geminate DEGEMINATION: a doubled consonant collapses word-finally (Ħadd→ħat, att→at) — but
     *  a MEDIAL geminate is kept (attard→attart, qattus→ʔattus). */
    private static void DegeminateFinal(List<string> toks)
    {
        var n = toks.Count;
        if (n >= 2 && toks[n - 1] == toks[n - 2] && !VOWELS.Contains(toks[n - 1])) toks.RemoveAt(n - 1);
    }

    /** A geminated AFFRICATE surfaces as its STOP plus the affricate (⟨ġġ⟩→[dd͡ʒ], ⟨ċċ⟩→[tt͡ʃ], ⟨zz⟩→[tt͡s]). */
    private static readonly IReadOnlyDictionary<string, string> AFFRICATE_STOP =
        new Dictionary<string, string>(StringComparer.Ordinal)
            { ["d͡ʒ"] = "d", ["t͡ʃ"] = "t", ["t͡s"] = "t" };

    private static void GeminateAffricate(List<string> toks)
    {
        for (var i = 0; i < toks.Count - 1; i++)
            if (AFFRICATE_STOP.TryGetValue(toks[i], out var stop) && toks[i] == toks[i + 1]) toks[i] = stop;
    }

    /** ⟨n⟩ → m before a labial (b/p/m): ġenb→d͡ʒɛmp. */
    private static void LabialNasal(List<string> toks)
    {
        for (var i = 0; i < toks.Count - 1; i++)
            if (toks[i] == "n" && (toks[i + 1] == "b" || toks[i + 1] == "p" || toks[i + 1] == "m"))
                toks[i] = "m";
    }

    /** Regressive voicing assimilation over obstruent clusters plus word-final devoicing (ħobż→ħɔps). */
    private static void ApplyVoicing(List<string> toks)
    {
        var last = toks.Count - 1;
        if (last >= 0 && DEVOICE.TryGetValue(toks[last], out var dv)) toks[last] = dv; // word-final devoicing
        for (var k = toks.Count - 2; k >= 0; k--)
        {
            var b = toks[k];
            var nb = toks[k + 1];
            if (DEVOICE.TryGetValue(b, out var d2) && VOICELESS.Contains(nb)) toks[k] = d2;
            else if (VOICE.TryGetValue(b, out var v2) && DEVOICE.ContainsKey(nb)) toks[k] = v2;
        }
    }

    /** Phonemize a single Maltese word to canonical IPA (segmental; length and stress folded/deferred). */
    public static string PhonemizeWord(string word)
    {
        var toks = Scan(word);
        CollapseVowels(toks);
        DegeminateFinal(toks);
        GeminateAffricate(toks);
        LabialNasal(toks);
        ApplyVoicing(toks);
        return string.Concat(toks);
    }

    /** A word (Maltese Latin letters incl. ċ ġ ħ ż, and the apostrophe) / number / punctuation token. */
    private static readonly JsRe TOKEN =
        JsRegex.Compile($"({HostWord.HostWordRun(new[] { "Latin" }, "'")})|(\\d+)|([.!?…,;:])", "gu");

    /**
     * This language's OWN inventory. ⚠ TWO DIFFERENT QUESTIONS, KEPT APART: the TOKEN class above decides
     * where the SCRIPT boundary falls (routing), while this one decides whether the g2p has rules for these
     * letters. A token this class REJECTS carries a letter the language does not use — a foreign name.
     *
     * ⚠ ⟨à ò⟩ JOIN ⟨è ì ù⟩ HERE FOR TRUTH, NOT FOR BEHAVIOUR (#1140). The note that used to sit here said the
     * opposite — that they were "deliberately absent" because the g2p "drops them outright" — and both
     * halves were false: `maltese.jsonc` declares all five grave vowels and this class carried only three.
     * Measured before changing it, the graves read the SAME QUALITY as their plain counterparts, so the fold
     * was reaching the right answer by the wrong route and listing them moves no output at all. Listed
     * anyway, because the class is a claim about the g2p and the claim was false.
     */
    private const string NATIVE_CLASS = "[a-zċġħżàèìòùA-ZĊĠĦŻÀÈÌÒÙ']";
    private static readonly Func<string, string> Nat = HostWord.MakeNativiser(NATIVE_CLASS, "u");

    public string Text(string input) =>
        // ⚠ TEXT NORMALIZATION FIRST — Normalize.cs rewrites grouping commas, decimal points, %, currency
        // signs, units, exponents, °C/°F, the minus, `&` and the Q.K./W.K. era markers into Maltese WORDS,
        // and it carries the shared symbol tier with them. Everything it emits is plain text, so it reaches
        // the g2p through the tokenizer below like any other word; nothing here writes a spelling into the
        // phoneme sink.
        Clauses.AssembleClauses(Normalize.NormalizeMaltese(input), TOKEN, (m, sink) =>
        {
            if (m.Groups[1].Success && m.Groups[1].Value.Length > 0)
            {
                sink.Emit(PhonemizeWord(Nat(m.Groups[1].Value)));
            }
            else if (m.Groups[2].Success && m.Groups[2].Value.Length > 0)
            {
                // ≤12 digits stays inside the attested range (< 10¹²); longer reads the raw digit string so
                // the Number() conversion cannot lose precision or go exponential.
                var tok = m.Groups[2].Value;
                var words = tok.Length <= 12
                    ? Numbers.NumberToWords(Js.Number(tok), NUM)
                    : Numbers.ReadDigits(tok, NUM);
                // "tnax-il" is one WORD whose hyphen the grapheme scan simply skips (→ ħdaxil), which is the
                // correct Maltese pronunciation of the linker; it is not a token boundary.
                foreach (var wd in words.Split(' ')) sink.Emit(PhonemizeWord(wd));
            }
            else if (m.Groups[3].Success && m.Groups[3].Value.Length > 0)
            {
                if (CLAUSE_MARK.TryGetValue(m.Groups[3].Value, out var mk) && mk.Length > 0) sink.Pause(mk);
            }
        });

    /** Build the Maltese phonemizer (grapheme g2p + silent-letter rules + devoicing). */
    public static ILanguage CreateMaltese() => new MaltesePhonemizer();

    internal static void RegisterSelf() => Registry.Register("maltese", CreateMaltese);
}
